dnadoh/synthesize.py

```diff
@@ -0,0 +1,335 @@
+    the other probabilities in `SNP_PROBS`.
+    """
+    candidates = _other_bases(reference, loc)
+    bases = [reference[loc]] + random.sample(candidates, k=len(candidates))
+    choices = random.choices(bases, weights=SNP_PROBS, k=len(genomes))
+    result = []
+    for (i, g) in enumerate(genomes):
+        result.append(g[:loc] + choices[i] + g[loc + 1 :])
+    return result
+
+
+def _mutate_other(genome, max_num_mutations, locations):
+    """Introduce up to `max_num_mutations` at specified locations."""
+    num = min(max_num_mutations, len(genome))
+    for loc in random.sample(locations, k=num):
+        base = random.choice(_other_bases(genome, loc))
+        genome = genome[:loc] + base + genome[loc + 1 :]
+    return genome
+
+
+# --------------------------------------------------------------------------------------
+# People
+# --------------------------------------------------------------------------------------
+
+
+class Person(BaseModel):
+    """An individual person.
+
+    Values marked `Optional` are filled in one at a time.
+    """
+
+    # Genome.
+    genome: str
+
+    # Age in years.
+    age: Optional[int] = None
+
+    # Genetic sex {"F", "M", "O"}.
+    gsex: Optional[str] = None
+
+    # Weight in kg.
+    weight: Optional[float] = None
+
+
+class PersonGenerator:
+    """Generate a person given a set of parameters."""
+
+    # Minimum and maximum ages in years.
+    MIN_AGE = 25
+    MAX_AGE = 75
+
+    # Genetic sex and probability of each.
+    GSEX = ("F", "M", "O")
+    GSEX_PROBS = (0.49, 0.48, 0.03)
+
+    # Mean weight in kg by genetic sex and relative standard deviation of weights.
+    WEIGHT_MEANS = {"F": 80.0, "M": 87.0, "O": 84.0}
+    WEIGHT_RSD = 0.12
+
+    def __init__(self):
+        """Construct generator."""
+
+    def make(self, reference, individual):
+        """Generate a new random person."""
+        person = Person(genome=individual)
+        for meth in (self.make_age, self.make_gsex, self.make_weight):
+            meth(person, reference, individual)
+        return person
+
+    def make_age(self, person, reference, individual):
+        """Generate a random age.
+
+        Ages are selected uniformly within bounds.
+        """
+        person.age = random.randint(self.MIN_AGE, self.MAX_AGE)
+
+    def make_gsex(self, person, reference, individual):
+        """Generate a random genetic sex.
+
+        Genetic sexes are chosen using weighted probabilities.
+        """
+        person.gsex = random.choices(self.GSEX, weights=self.GSEX_PROBS)[0]
+
+    def make_weight(self, person, reference, individual):
+        """Generate a random weight.
+
+        Weights are chosen from a normal distribution based on genetic sex.
+        """
+        assert person.gsex is not None
+        mean = self.WEIGHT_MEANS[person.gsex]
+        std_dev = mean * self.WEIGHT_RSD
+        person.weight = _truncate(random.gauss(mean, std_dev), 1)
```

Comment:
is there no library that can do this? what about rounding instead of truncating?